fl4health/clients/mr_mtl_client.py

```diff
@@ -0,0 +1,188 @@
+        # Need to setup the init global model here as well. It should be the same architecture as the model so
+        # we reuse the get_model call. We explicitly send the model to the desired device. This is idempotent.
+        self.init_global_model = self.get_model(config).to(self.device)
+        # The rest of the setup is the same
+        super().setup_client(config)
+
+    def get_mr_drift_loss(self) -> torch.Tensor:
+        """
+        Compute the L2 inner product between the initial global weights for the round and the current local model
+            weights. This loss function is added to the loss function for the local model when back propagating.
+
+        Returns:
+            torch.Tensor: Returns the L2 inner product between the initial global weights of the round and the current
+                local model weights.
+        """
+        assert self.init_global_model is not None and self.model is not None and self.lam is not None
+        # Using parameters to ensure the same ordering as exchange
+        local_model_weights = [layer_weights for layer_weights in self.model.parameters()]
+        # Detach the weights to ensure we don't compute gradients with respect to the tensors
+        initial_global_weights = [layer_weights.detach() for layer_weights in self.init_global_model.parameters()]
+
+        assert len(initial_global_weights) == len(local_model_weights)
+        assert len(initial_global_weights) > 0
+
+        layer_inner_products: List[torch.Tensor] = [
+            torch.pow(torch.linalg.norm(initial_layer_weights - iteration_layer_weights), 2.0)
+            for initial_layer_weights, iteration_layer_weights in zip(initial_global_weights, local_model_weights)
+        ]
+
+        # network l2 inner product tensor weighted by lambda
+        return (self.lam / 2.0) * torch.stack(layer_inner_products).sum()
+
+    def get_parameters(self, config: Config) -> NDArrays:
+        """
+        For MR-MTL, we transfer the LOCAL model weights to the server to be aggregated and set as INITIAL GLOBAL model
+        weightson client side.
+
+        Args:
+            config (Config): The config is sent by the FL server to allow for customization in the function if desired.
+
+        Returns:
+            NDArrays: GLOBAL model weights to be sent to the server for aggregation
+        """
+        assert self.model is not None and self.parameter_exchanger is not None
+        return self.parameter_exchanger.push_parameters(self.model, config=config)
+
+    def set_parameters(self, parameters: NDArrays, config: Config, fitting_round: bool) -> None:
+        """
+        The parameters being pass are to be routed to the initial global model to be used in a penalty term in
+        training the local model.
+
+        Args:
+            parameters (NDArrays): Parameters have information about model state to be added to the relevant client
+                model (global model for all but the first step of MR-MTL)
```

Comment:
So it looks like, based on the implementation below, we actually only ever set the global model (even if its the first step). Based on algorithm A.1 from the paper, I think that's fine. That is, each client starts with a random initial model but the same global model. However, we should talk about that in the description for this method if that's how we're doing it. That is, make that behavior explicit in the comments.

Reply:
Good idea, I added a comment about this to the description.